src/reactjs_jsx_ppx_v4.ml

```diff
@@ -0,0 +1,1455 @@
+open Ast_helper
+open Ast_mapper
+open Asttypes
+open Parsetree
+open Longident
+
+module PropsMap = Map.Make(String)
+
+type domProp = {
+  jsString : string option;
+  typeString : Parsetree.core_type;
+}
+
+let nolabel = Nolabel
+let labelled str = Labelled str
+let optional str = Optional str
+
+
+let mkPropConst ident = Typ.constr {loc = Location.none; txt = ident} []
+let mkSingleVariant rf = Rtag ({loc = Location.none; txt = rf}, [], true, [])
+let mkPropVariant row_fields = Typ.variant (List.map mkSingleVariant row_fields) Closed None
+
+let allowedProps = [
+("children", {jsString = None; typeString = mkPropConst (Ldot (Lident "React", "element"))});
+("key", {jsString = None; typeString = mkPropConst (Lident "string")});
+("ref", {jsString = None; typeString = mkPropConst (Ldot (Lident "ReactDOM", "domRef"))});
+("ariaCurrent", {jsString = Some("aria-current"); typeString = mkPropVariant ["page"; "step"; "location"; "date"; "time"; "true"; "false"]});
+("ariaDetails", {jsString = Some("aria-details"); typeString = mkPropConst (Lident "string")});
+("ariaDisabled", {jsString = Some("aria-disabled"); typeString = mkPropConst (Lident "bool")});
+("ariaHidden", {jsString = Some("aria-hidden"); typeString = mkPropConst (Lident "bool")});
+("ariaInvalid", {jsString = Some("aria-invalid"); typeString = mkPropVariant ["grammar"; "false"; "spelling"; "true"]});
+("ariaKeyshortcuts", {jsString = Some("aria-keyshortcuts"); typeString = mkPropConst (Lident "string")});
+("ariaLabel", {jsString = Some("aria-label"); typeString = mkPropConst (Lident "string")});
+("ariaRoledescription", {jsString = Some("aria-roledescription"); typeString = mkPropConst (Lident "string")});
+("ariaAutocomplete", {jsString = Some("aria-autocomplete"); typeString = mkPropVariant ["inline"; "list"; "both"; "none"]});
+("ariaChecked", {jsString = Some("aria-checked"); typeString = mkPropVariant ["true"; "false" ;"mixed"]});
+("ariaExpanded", {jsString = Some("aria-expanded"); typeString = mkPropConst (Lident "bool")});
+("ariaHaspopup", {jsString = Some("aria-haspopup"); typeString = mkPropVariant ["false"; "true"; "menu"; "listbox"; "tree"; "grid"; "dialog"]});
+("ariaLevel", {jsString = Some("aria-level"); typeString = mkPropConst (Lident "int")});
+("ariaModal", {jsString = Some("aria-modal"); typeString = mkPropConst (Lident "bool")});
+("ariaMultiline", {jsString = Some("aria-multiline"); typeString = mkPropConst (Lident "bool")});
+("ariaMultiselectable", {jsString = Some("aria-multiselectable"); typeString = mkPropConst (Lident "bool")});
+("ariaOrientation", {jsString = Some("aria-orientation"); typeString = mkPropVariant ["horizontal"; "vertical"; "undefined"]});
+("ariaPlaceholder", {jsString = Some("aria-placeholder"); typeString = mkPropConst (Lident "string")});
+("ariaPressed", {jsString = Some("aria-pressed"); typeString = mkPropVariant ["true"; "false" ;"mixed"]});
```

Comment:
Similar as the other - does explicit "undefined" have different behavior from not providing ariaPressed at all? Much smaller downside to including here though.

Reply:
Yes, in this case we already *lost* the `bool` because of the `mixed` variant. If we don't need to express the `undefined` variant as discussed before, it is just a matter of choice. ☺️